README.md
# SCRU128: Sortable, Clock and Random number-based Unique identifier

[![GitHub tag](https://img.shields.io/github/v/tag/scru128/swift-scru128)](https://github.com/scru128/swift-scru128)
[![License](https://img.shields.io/github/license/scru128/swift-scru128)](https://github.com/scru128/swift-scru128/blob/main/LICENSE)
[![Swift Compatibility](https://img.shields.io/endpoint?url=https%3A%2F%2Fswiftpackageindex.com%2Fapi%2Fpackages%2Fscru128%2Fswift-scru128%2Fbadge%3Ftype%3Dswift-versions)](https://swiftpackageindex.com/scru128/swift-scru128)
[![Platform Compatibility](https://img.shields.io/endpoint?url=https%3A%2F%2Fswiftpackageindex.com%2Fapi%2Fpackages%2Fscru128%2Fswift-scru128%2Fbadge%3Ftype%3Dplatforms)](https://swiftpackageindex.com/scru128/swift-scru128)

SCRU128 ID is yet another attempt to supersede [UUID] for the users who need
decentralized, globally unique time-ordered identifiers. SCRU128 is inspired by
[ULID] and [KSUID] and has the following features:

- 128-bit unsigned integer type
- Sortable by generation time (as integer and as text)
- 25-digit case-insensitive textual representation (Base36)
- 48-bit millisecond Unix timestamp that ensures useful life until year 10889
- Up to 281 trillion time-ordered but unpredictable unique IDs per millisecond
- 80-bit three-layer randomness for global uniqueness

```swift
import Scru128

// generate a new identifier object
let x = scru128()
print(x)  // e.g., "036z951mhjikzik2gsl81gr7l"
print(x.byteArray)  // as a 128-bit unsigned integer in big-endian byte array

// generate a textual representation directly
print(scru128String())  // e.g., "036z951mhzx67t63mq9xe6q0j"
```

See [SCRU128 Specification] for details.

[UUID]: https://en.wikipedia.org/wiki/Universally_unique_identifier
[ULID]: https://github.com/ulid/spec
[KSUID]: https://github.com/segmentio/ksuid
[SCRU128 Specification]: https://github.com/scru128/spec

## Add swift-scru128 as a package dependency

To add this library to your Xcode project as a dependency, select **File** >
**Add Packages** and enter the package URL:
https://github.com/scru128/swift-scru128

To use this library in a SwiftPM project, add the following line to the
dependencies in your Package.swift file:

```swift
.package(url: "https://github.com/scru128/swift-scru128", from: "<version>"),
```

And, include `Scru128` as a dependency for your target:

```swift
.target(
  name: "<target>",
  dependencies: [.product(name: "Scru128", package: "swift-scru128")]
)
```

## License

Licensed under the Apache License, Version 2.0.

## See also

- [swift-scru128 - Swift Package Index](https://swiftpackageindex.com/scru128/swift-scru128)


CHANGELOG.md
# Changelog

## v3.0.2 - 2023-09-18

### Fixed

- Mishandling of `UnsafeMutableBufferPointer` in `Scru128Id#description`

### Maintenance

- Refactored private methods and error messages
- Improved documentation about generator's clock rollback behavior
- Updated README

## v3.0.1 - 2023-07-30

Most notably, v3 switches the letter case of generated IDs from uppercase (e.g.,
"036Z951MHJIKZIK2GSL81GR7L") to lowercase (e.g., "036z951mhjikzik2gsl81gr7l"),
though it is technically not supposed to break existing code because SCRU128 is
a case-insensitive scheme. Other changes include the removal of deprecated APIs.

### Removed

- Deprecated items:
  - `Scru128Generator#generateCore()`
  - `Scru128Generator#lastStatus` and `Scru128Generator.Status`

### Changed

- Letter case of generated IDs from uppercase to lowercase
- Return type of `Scru128Id#bytes` property from `[UInt8]` to tuple of 16
  `UInt8` byte values, as a result of change in internal representation of
  `Scru128Id`
- `Scru128Generator` to `Scru128Generator<R: RandomNumberGenerator>` to
  substitute static generics for dynamic existential type
- Edge case behavior of generator functions' rollback allowance handling

### Added

- `Scru128Id#byteArray` to emulate old `Scru128Id#bytes` property

## v2.4.6 - 2023-07-30

### Added

- `Scru128Id#byteArray` as synonym for deprecated `bytes`
- `Scru128Id` initializer that receives tuple of 16 byte values

### Deprecated

- `Scru128Id#bytes` to help migration to v3

## v2.4.5 - 2023-07-29

### Fixed

- Test case that could not compile with Xcode 13.4

## v2.4.4 - 2023-06-21

### Maintenance

- Improved test cases

## v2.4.3 - 2023-06-05

### Fixed

- multi-threaded test cases by adding proper @available attribute

## v2.4.2 - 2023-06-04

### Added

- `Sendable` protocol conformance to `Scru128Id`

### Maintenance

- Upgraded minimum Swift version to 5.6
- Rewrote multi-threaded test cases using Swift Concurrency
- Fixed README

## v2.4.1 - 2023-04-08

### Maintenance

- Tweaked docs and tests

## v2.4.0 - 2023-03-22

### Added

- `generateOrAbort()` and `generateOrAbortCore()` to `Scru128Generator`
  (formerly named as `generateNoRewind()` and `generateCoreNoRewind()`)
- `Scru128Generator#generateOrResetCore()`

### Deprecated

- `Scru128Generator#generateCore()`
- `Scru128Generator#lastStatus` and `Scru128Generator.Status`

## v2.3.1 - 2023-03-19

### Added

- `generateNoRewind()` and `generateCoreNoRewind(timestamp:rollbackAllowance:)`
  to `Scru128Generator` (experimental)

### Maintenance

- Improved documentation about generator method flavors

## v2.3.0 - 2023-02-13

### Fixed

- `Decodable` implementation of `Scru128Id` to throw error on failure instead of
  raising runtime panic

### Changed

- `Decodable` behavior so it tries to parse byte array as well as string

## v2.2.0 - 2022-12-23

### Added

- Sequence and IteratorProtocol implementations to `Scru128Generator` to make it
  work as infinite iterator

## v2.1.2 - 2022-06-11

### Fixed

- `generateCore()` to update `counter_hi` when `timestamp` passed < 1000

## v2.1.1 - 2022-05-23

### Fixed

- `generateCore()` to reject zero as `timestamp` value

## v2.1.0 - 2022-05-22

### Added

- `generateCore()` and `lastStatus` to `Scru128Generator`

## v2.0.0 - 2022-05-01

### Changed

- Textual representation: 26-digit Base32 -> 25-digit Base36
- Field structure: { `timestamp`: 44 bits, `counter`: 28 bits, `per_sec_random`:
  24 bits, `per_gen_random`: 32 bits } -> { `timestamp`: 48 bits, `counter_hi`:
  24 bits, `counter_lo`: 24 bits, `entropy`: 32 bits }
- Timestamp epoch: 2020-01-01 00:00:00.000 UTC -> 1970-01-01 00:00:00.000 UTC
- Counter overflow handling: stall generator -> increment timestamp

### Removed

- `setScru128Logger()` as counter overflow is no longer likely to occur
- `Scru128Id#counter`, `Scru128Id#perSecRandom`, `Scru128Id#perGenRandom`

### Added

- `Scru128Id#counterHi`, `Scru128Id#counterLo`, `Scru128Id#entropy`

## v1.0.0 - 2022-01-03

- Initial stable release


Tests/Scru128Tests/Scru128Tests.swift
import XCTest

@testable import Scru128

final class Scru128Tests: XCTestCase {
  static let samples: [String] = (0..<100_000).map { _ in scru128String() }

  /// Generates 25-digit canonical string
  func testFormat() throws {
    for e in Self.samples {
      XCTAssertNotNil(e.range(of: "^[0-9a-z]{25}$", options: .regularExpression))
    }
  }

  /// Generates 100k identifiers without collision
  func testUniqueness() throws {
    let set = Set(Self.samples)
    XCTAssertEqual(set.count, Self.samples.count)
  }

  /// Generates sortable string representation by creation time
  func testOrder() throws {
    for i in 1..<Self.samples.count {
      XCTAssertLessThan(Self.samples[i - 1], Self.samples[i])
    }
  }

  /// Encodes up-to-date timestamp
  func testTimestamp() throws {
    let g = Scru128Generator()
    for _ in 0..<10_000 {
      let tsNow = Int64(Date().timeIntervalSince1970 * 1_000)
      let timestamp = Int64(g.generate().timestamp)
      XCTAssertLessThan(abs(tsNow - timestamp), 16)
    }
  }

  /// Encodes unique sortable tuple of timestamp and counters
  func testTimestampAndCounters() throws {
    var prev = Scru128Id(Self.samples[0])!
    for i in 1..<Self.samples.count {
      let curr = Scru128Id(Self.samples[i])!
      XCTAssertTrue(
        prev.timestamp < curr.timestamp
          || (prev.timestamp == curr.timestamp && prev.counterHi < curr.counterHi)
          || (prev.timestamp == curr.timestamp && prev.counterHi == curr.counterHi
            && prev.counterLo < curr.counterLo)
      )
      prev = curr
    }
  }

  /// Generates no IDs sharing same timestamp and counters under multithreading
  @available(iOS 13.0, macOS 10.15, macCatalyst 13.0, tvOS 13.0, watchOS 6.0, *)
  func testThreading() async throws {
    let results = await withTaskGroup(of: [Scru128Id].self) { (group) -> [Scru128Id] in
      for _ in 0..<4 {
        group.addTask { (0..<10_000).map { _ in scru128() } }
      }

      var results: [Scru128Id] = []
      for await xs in group {
        results.append(contentsOf: xs)
      }
      return results
    }

    let set = Set<String>(results.map { "\($0.timestamp)-\($0.counterHi)-\($0.counterLo)" })
    XCTAssertEqual(set.count, results.count)
  }
}


Tests/Scru128Tests/Scru128IdTests.swift
import XCTest

@testable import Scru128

let maxUint48: UInt64 = (1 << 48) - 1
let maxUint24: UInt32 = (1 << 24) - 1
let maxUint32: UInt32 = UInt32.max

final class Scru128IdTests: XCTestCase {
  /// Encodes and decodes prepared cases correctly
  func testEncodeDecode() throws {
    let cases: [((UInt64, UInt32, UInt32, UInt32), String)] = [
      ((0, 0, 0, 0), "0000000000000000000000000"),
      ((maxUint48, 0, 0, 0), "F5LXX1ZZ5K6TP71GEEH2DB7K0"),
      ((maxUint48, 0, 0, 0), "f5lxx1zz5k6tp71geeh2db7k0"),
      ((0, maxUint24, 0, 0), "0000000005GV2R2KJWR7N8XS0"),
      ((0, maxUint24, 0, 0), "0000000005gv2r2kjwr7n8xs0"),
      ((0, 0, maxUint24, 0), "00000000000000JPIA7QL4HS0"),
      ((0, 0, maxUint24, 0), "00000000000000jpia7ql4hs0"),
      ((0, 0, 0, maxUint32), "0000000000000000001Z141Z3"),
      ((0, 0, 0, maxUint32), "0000000000000000001z141z3"),
      ((maxUint48, maxUint24, maxUint24, maxUint32), "F5LXX1ZZ5PNORYNQGLHZMSP33"),
      ((maxUint48, maxUint24, maxUint24, maxUint32), "f5lxx1zz5pnorynqglhzmsp33"),
    ]

    for e in cases {
      let fromFields = Scru128Id(e.0.0, e.0.1, e.0.2, e.0.3)
      let fromString = Scru128Id(e.1)!

      XCTAssertEqual(fromFields, fromString)
      XCTAssertEqual(fromFields.timestamp, e.0.0)
      XCTAssertEqual(fromString.timestamp, e.0.0)
      XCTAssertEqual(fromFields.counterHi, e.0.1)
      XCTAssertEqual(fromString.counterHi, e.0.1)
      XCTAssertEqual(fromFields.counterLo, e.0.2)
      XCTAssertEqual(fromString.counterLo, e.0.2)
      XCTAssertEqual(fromFields.entropy, e.0.3)
      XCTAssertEqual(fromString.entropy, e.0.3)
      XCTAssertEqual(fromFields.description, e.1.lowercased())
      XCTAssertEqual(fromString.description, e.1.lowercased())
    }
  }

  /// Returns error if an invalid string representation is supplied
  func testStringValidation() throws {
    let cases = [
      "",
      " 036z8puq4tsxsigk6o19y164q",
      "036z8puq54qny1vq3hcbrkweb ",
      " 036z8puq54qny1vq3helivwax ",
      "+036z8puq54qny1vq3hfcv3ss0",
      "-036z8puq54qny1vq3hhy8u1ch",
      "+36z8puq54qny1vq3hjq48d9p",
      "-36z8puq5a7j0ti08oz6zdrdy",
      "036z8puq5a7j0t_08p2cdz28v",
      "036z8pu-5a7j0ti08p3ol8ool",
      "036z8puq5a7j0ti08p4j 6cya",
      "f5lxx1zz5pnorynqglhzmsp34",
      "zzzzzzzzzzzzzzzzzzzzzzzzz",
      "039o\tvvklfmqlqe7fzllz7c7t",
      "039onvvklfmqlq漢字fgvd1",
      "039onvvkl🤣qe7fzr2hdoqu",
      "頭onvvklfmqlqe7fzrhtgcfz",
      "039onvvklfmqlqe7fztft5尾",
      "039漢字a52xp4bvf4sn94e09cja",
      "039ooa52xp4bv😘sn97642mwl",
    ]

    for e in cases {
      XCTAssertNil(Scru128Id(e))
    }
  }

  /// Has symmetric converters from/to various values
  func testSymmetricConverters() throws {
    let encoder = JSONEncoder()
    let decoder = JSONDecoder()

    var cases = [
      Scru128Id(0, 0, 0, 0),
      Scru128Id(maxUint48, 0, 0, 0),
      Scru128Id(0, maxUint24, 0, 0),
      Scru128Id(0, 0, maxUint24, 0),
      Scru128Id(0, 0, 0, maxUint32),
      Scru128Id(maxUint48, maxUint24, maxUint24, maxUint32),
    ]

    let g = Scru128Generator()
    for _ in 0..<1_000 {
      cases.append(g.generate())
    }

    for e in cases {
      XCTAssertEqual(Scru128Id(e.description)!, e)
      XCTAssertEqual(Scru128Id(e.bytes), e)
      XCTAssertEqual(Scru128Id(e.byteArray), e)
      XCTAssertEqual(Scru128Id(e.timestamp, e.counterHi, e.counterLo, e.entropy), e)
      XCTAssertEqual(try decoder.decode(Scru128Id.self, from: try encoder.encode(e)), e)

      let tp = e.bytes
      let ar = e.byteArray
      XCTAssertEqual(
        Scru128Id(
          (
            ar[0], ar[1], ar[2], ar[3], ar[4], ar[5], ar[6], ar[7],
            ar[8], ar[9], ar[10], ar[11], ar[12], ar[13], ar[14], ar[15]
          )
        ), e)
      XCTAssertEqual(
        Scru128Id([
          tp.0, tp.1, tp.2, tp.3, tp.4, tp.5, tp.6, tp.7,
          tp.8, tp.9, tp.10, tp.11, tp.12, tp.13, tp.14, tp.15,
        ]), e)
    }
  }

  /// Supports comparison operators
  func testComparisonOperators() throws {
    var ordered = [
      Scru128Id(0, 0, 0, 0),
      Scru128Id(0, 0, 0, 1),
      Scru128Id(0, 0, 0, maxUint32),
      Scru128Id(0, 0, 1, 0),
      Scru128Id(0, 0, maxUint24, 0),
      Scru128Id(0, 1, 0, 0),
      Scru128Id(0, maxUint24, 0, 0),
      Scru128Id(1, 0, 0, 0),
      Scru128Id(2, 0, 0, 0),
    ]

    let g = Scru128Generator()
    for _ in 0..<1_000 {
      ordered.append(g.generate())
    }

    var prev = ordered.removeFirst()
    for curr in ordered {
      XCTAssertNotEqual(curr, prev)
      XCTAssertNotEqual(prev, curr)
      XCTAssertNotEqual(curr.byteArray, prev.byteArray)
      XCTAssertNotEqual(curr.hashValue, prev.hashValue)
      XCTAssertGreaterThan(curr, prev)
      XCTAssertGreaterThanOrEqual(curr, prev)
      XCTAssertLessThan(prev, curr)
      XCTAssertLessThanOrEqual(prev, curr)

      let clone = curr
      XCTAssertEqual(curr, clone)
      XCTAssertEqual(clone, curr)
      XCTAssertEqual(curr.byteArray, clone.byteArray)
      XCTAssertEqual(curr.hashValue, clone.hashValue)
      XCTAssertGreaterThanOrEqual(curr, clone)
      XCTAssertGreaterThanOrEqual(clone, curr)
      XCTAssertLessThanOrEqual(curr, clone)
      XCTAssertLessThanOrEqual(clone, curr)

      prev = curr
    }
  }

  /// Serializes and deserializes an object using the canonical string representation
  func testSerializedForm() throws {
    let encoder = JSONEncoder()
    let decoder = JSONDecoder()

    let g = Scru128Generator()
    for _ in 0..<1_000 {
      let obj = g.generate()
      let strJson = "\"\(obj)\""
      XCTAssertEqual(String(data: try encoder.encode(obj), encoding: .utf8), strJson)
      XCTAssertEqual(try decoder.decode(Scru128Id.self, from: strJson.data(using: .utf8)!), obj)
    }
  }

  /// Decodes an object from string and binary representations
  func testDecodable() throws {
    let encoder = JSONEncoder()
    let decoder = JSONDecoder()

    let g = Scru128Generator()
    for _ in 0..<1_000 {
      let obj = g.generate()
      let desc = obj.description
      let asStr = try encoder.encode(desc)
      let asBinBytes = try encoder.encode(obj.byteArray)
      let asStrBytes = try encoder.encode([UInt8](desc.utf8))

      XCTAssertEqual(try decoder.decode(Scru128Id.self, from: asStr), obj)
      XCTAssertEqual(try decoder.decode(Scru128Id.self, from: asBinBytes), obj)
      XCTAssertEqual(try decoder.decode(Scru128Id.self, from: asStrBytes), obj)
    }
  }
}


Tests/Scru128Tests/Scru128GeneratorTests.swift
import XCTest

@testable import Scru128

final class Scru128GeneratorGenerateOrResetTests: XCTestCase {
  /// Generates increasing IDs even with decreasing or constant timestamp
  func testDecreasingOrConstantTimestamp() throws {
    let ts: UInt64 = 0x0123_4567_89ab
    let g = Scru128Generator()

    var prev = g.generateOrResetCore(timestamp: ts, rollbackAllowance: 10_000)
    XCTAssertEqual(prev.timestamp, ts)

    for i in UInt64(0)..<100_000 {
      let curr = g.generateOrResetCore(timestamp: ts - min(9_999, i), rollbackAllowance: 10_000)
      XCTAssertLessThan(prev, curr)
      prev = curr
    }
    XCTAssertGreaterThanOrEqual(prev.timestamp, ts)
  }

  /// Breaks increasing order of IDs if timestamp goes backwards a lot
  func testTimestampRollback() throws {
    let ts: UInt64 = 0x0123_4567_89ab
    let g = Scru128Generator()

    var prev = g.generateOrResetCore(timestamp: ts, rollbackAllowance: 10_000)
    XCTAssertEqual(prev.timestamp, ts)

    var curr = g.generateOrResetCore(timestamp: ts - 10_000, rollbackAllowance: 10_000)
    XCTAssertLessThan(prev, curr)

    prev = curr
    curr = g.generateOrResetCore(timestamp: ts - 10_001, rollbackAllowance: 10_000)
    XCTAssertGreaterThan(prev, curr)
    XCTAssertEqual(curr.timestamp, ts - 10_001)

    prev = curr
    curr = g.generateOrResetCore(timestamp: ts - 10_002, rollbackAllowance: 10_000)
    XCTAssertLessThan(prev, curr)
  }
}

final class Scru128GeneratorGenerateOrAbortTests: XCTestCase {
  /// Generates increasing IDs even with decreasing or constant timestamp
  func testDecreasingOrConstantTimestamp() throws {
    let ts: UInt64 = 0x0123_4567_89ab
    let g = Scru128Generator()

    var prev = g.generateOrAbortCore(timestamp: ts, rollbackAllowance: 10_000)!
    XCTAssertEqual(prev.timestamp, ts)

    for i in UInt64(0)..<100_000 {
      let curr = g.generateOrAbortCore(timestamp: ts - min(9_999, i), rollbackAllowance: 10_000)!
      XCTAssertLessThan(prev, curr)
      prev = curr
    }
    XCTAssertGreaterThanOrEqual(prev.timestamp, ts)
  }

  /// Returns nil if timestamp goes backwards a lot
  func testTimestampRollback() throws {
    let ts: UInt64 = 0x0123_4567_89ab
    let g = Scru128Generator()

    let prev = g.generateOrAbortCore(timestamp: ts, rollbackAllowance: 10_000)!
    XCTAssertEqual(prev.timestamp, ts)

    var curr = g.generateOrAbortCore(timestamp: ts - 10_000, rollbackAllowance: 10_000)
    XCTAssertLessThan(prev, curr!)

    curr = g.generateOrAbortCore(timestamp: ts - 10_001, rollbackAllowance: 10_000)
    XCTAssertNil(curr)

    curr = g.generateOrAbortCore(timestamp: ts - 10_002, rollbackAllowance: 10_000)
    XCTAssertNil(curr)
  }
}

final class Scru128GeneratorTests: XCTestCase {
  /// Is iterable with for-in loop
  func testSequenceImplementation() throws {
    var i = 0
    for e in Scru128Generator() {
      XCTAssertGreaterThan(e.timestamp, 0)
      i += 1
      if i > 100 {
        break
      }
    }
    XCTAssertEqual(i, 101)
  }
}


Sources/Scru128/Scru128Id.swift
/// Represents a SCRU128 ID and provides converters and comparison operators.
public struct Scru128Id: LosslessStringConvertible {
  /// Returns a 16-byte byte tuple containing the 128-bit unsigned integer representation in the
  /// big-endian (network) byte order.
  public let bytes:
    (
      UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8,
      UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8
    )

  /// Creates an object from a byte tuple that represents a 128-bit unsigned integer.
  ///
  /// - Parameter bytes: A 16-byte byte array that represents a 128-bit unsigned integer in the
  ///                    big-endian (network) byte order.
  public init(
    _ bytes: (
      UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8,
      UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8
    )
  ) {
    self.bytes = bytes
  }

  /// Creates an object from a byte array that represents a 128-bit unsigned integer.
  ///
  /// - Parameter byteArray: A 16-byte byte array that represents a 128-bit unsigned integer in the
  ///                        big-endian (network) byte order.
  /// - Precondition: The byte length of the argument must be 16.
  public init(_ byteArray: [UInt8]) {
    precondition(byteArray.count == 16)
    bytes = (
      byteArray[0], byteArray[1], byteArray[2], byteArray[3],
      byteArray[4], byteArray[5], byteArray[6], byteArray[7],
      byteArray[8], byteArray[9], byteArray[10], byteArray[11],
      byteArray[12], byteArray[13], byteArray[14], byteArray[15]
    )
  }

  /// Creates an object from field values.
  ///
  /// - Parameters:
  ///   - timestamp: A 48-bit `timestamp` field value.
  ///   - counterHi: A 24-bit `counter_hi` field value.
  ///   - counterLo: A 24-bit `counter_lo` field value.
  ///   - entropy: A 32-bit `entropy` field value.
  /// - Precondition: Each argument must be within the value range of the field.
  public init(
    _ timestamp: UInt64, _ counterHi: UInt32, _ counterLo: UInt32, _ entropy: UInt32
  ) {
    precondition(timestamp <= maxTimestamp)
    precondition(counterHi <= maxCounterHi)
    precondition(counterLo <= maxCounterLo)
    bytes = (
      UInt8(truncatingIfNeeded: timestamp >> 40),
      UInt8(truncatingIfNeeded: timestamp >> 32),
      UInt8(truncatingIfNeeded: timestamp >> 24),
      UInt8(truncatingIfNeeded: timestamp >> 16),
      UInt8(truncatingIfNeeded: timestamp >> 8),
      UInt8(truncatingIfNeeded: timestamp),
      UInt8(truncatingIfNeeded: counterHi >> 16),
      UInt8(truncatingIfNeeded: counterHi >> 8),
      UInt8(truncatingIfNeeded: counterHi),
      UInt8(truncatingIfNeeded: counterLo >> 16),
      UInt8(truncatingIfNeeded: counterLo >> 8),
      UInt8(truncatingIfNeeded: counterLo),
      UInt8(truncatingIfNeeded: entropy >> 24),
      UInt8(truncatingIfNeeded: entropy >> 16),
      UInt8(truncatingIfNeeded: entropy >> 8),
      UInt8(truncatingIfNeeded: entropy)
    )
  }

  /// Creates an object from a 25-digit string representation.
  public init?(_ description: String) {
    try? self.init(parsing: description)
  }

  /// Builds the 16-byte big-endian byte array representation from a string.
  private init(parsing: String) throws {
    var description = parsing
    let byteArray = try description.withUTF8 {
      if $0.count != 25 {
        throw ParseError(message: "invalid length: \($0.count) bytes (expected 25)")
      }

      var src = [UInt8](repeating: 0, count: 25)
      for i in 0..<src.count {
        src[i] = decodeMap[Int($0[i])]
        if src[i] == 0xff {
          throw ParseError(message: "invalid digit at \(i)")
        }
      }

      var dst = [UInt8](repeating: 0, count: 16)
      var minIndex = 99  // any number greater than size of output array
      for i in stride(from: -5, to: 25, by: 10) {
        // implement Base36 using 10-digit words
        var carry: UInt64 = src[(i < 0 ? 0 : i)..<(i + 10)].reduce(0) { $0 * 36 + UInt64($1) }

        // iterate over output array from right to left while carry != 0 but at least up to place
        // already filled
        var j = dst.count - 1
        while carry > 0 || j > minIndex {
          if j < 0 {
            throw ParseError(message: "out of 128-bit value range")
          }
          carry += UInt64(dst[j]) * 3_656_158_440_062_976  // 36^10
          dst[j] = UInt8(truncatingIfNeeded: carry)
          carry = carry >> 8
          j -= 1
        }
        minIndex = j
      }
      return dst
    }
    self.init(byteArray)
  }

  /// An error parsing an invalid string representation of SCRU128 ID.
  private struct ParseError: Error, CustomStringConvertible {
    let message: String
    var description: String { "could not parse string as SCRU128 ID: \(message)" }
  }

  /// Returns a 16-byte byte array containing the 128-bit unsigned integer representation in the
  /// big-endian (network) byte order.
  public var byteArray: [UInt8] { (0..<16).map(byteAt) }

  /// Returns the 48-bit `timestamp` field value.
  public var timestamp: UInt64 { subUInt(0..<6) }

  /// Returns the 24-bit `counter_hi` field value.
  public var counterHi: UInt32 { subUInt(6..<9) }

  /// Returns the 24-bit `counter_lo` field value.
  public var counterLo: UInt32 { subUInt(9..<12) }

  /// Returns the 32-bit `entropy` field value.
  public var entropy: UInt32 { subUInt(12..<16) }

  /// Returns the 25-digit canonical string representation.
  public var description: String {
    func buildUtf8Bytes(_ dst: UnsafeMutableBufferPointer<UInt8>) -> Int {
      dst.initialize(repeating: 0)
      var minIndex = 99  // any number greater than size of output array
      for i in stride(from: -5, to: 16, by: 7) {
        // implement Base36 using 56-bit words
        var carry: UInt64 = subUInt((i < 0 ? 0 : i)..<(i + 7))

        // iterate over output array from right to left while carry != 0 but at least up to place
        // already filled
        var j = 24
        while carry > 0 || j > minIndex {
          carry += UInt64(dst[j]) << 56
          dst[j] = UInt8(truncatingIfNeeded: carry % 36)
          carry = carry / 36
          j -= 1
        }
        minIndex = j
      }

      for i in 0..<25 {
        dst[i] = digits[Int(dst[i])]
      }
      return 25
    }

    if #available(iOS 14.0, macOS 11.0, macCatalyst 14.0, tvOS 14.0, watchOS 7.0, *) {
      return String(unsafeUninitializedCapacity: 25, initializingUTF8With: buildUtf8Bytes)
    } else {
      return String(
        cString: [UInt8](unsafeUninitializedCapacity: 26) { $1 = buildUtf8Bytes($0) + 1 })
    }
  }

  /// Returns a part of `bytes` as an unsigned integer.
  private func subUInt<T: UnsignedInteger>(_ range: Range<Int>) -> T {
    range.reduce(0) { $0 << 8 | T(byteAt($1)) }
  }

  /// Returns the byte value of `self` at an index.
  private func byteAt(_ index: Int) -> UInt8 {
    precondition(0 <= index && index < 16)

    // implement binary search
    if index < 8 {
      if index < 4 {
        if index < 2 {
          return index < 1 ? bytes.0 : bytes.1
        } else {
          return index < 3 ? bytes.2 : bytes.3
        }
      } else {
        if index < 6 {
          return index < 5 ? bytes.4 : bytes.5
        } else {
          return index < 7 ? bytes.6 : bytes.7
        }
      }
    } else {
      if index < 12 {
        if index < 10 {
          return index < 9 ? bytes.8 : bytes.9
        } else {
          return index < 11 ? bytes.10 : bytes.11
        }
      } else {
        if index < 14 {
          return index < 13 ? bytes.12 : bytes.13
        } else {
          return index < 15 ? bytes.14 : bytes.15
        }
      }
    }
  }
}

/// Digit characters used in the Base36 notation.
private let digits = [UInt8]("0123456789abcdefghijklmnopqrstuvwxyz".utf8)

/// An O(1) map from ASCII code points to Base36 digit values.
private let decodeMap: [UInt8] = [
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18,
  0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18,
  0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
]

extension Scru128Id: Comparable, Hashable {
  public static func == (lhs: Scru128Id, rhs: Scru128Id) -> Bool {
    (0..<16).allSatisfy { lhs.byteAt($0) == rhs.byteAt($0) }
  }

  public static func < (lhs: Scru128Id, rhs: Scru128Id) -> Bool {
    for i in 0..<16 {
      let lft = lhs.byteAt(i)
      let rgt = rhs.byteAt(i)
      if lft != rgt {
        return lft < rgt
      }
    }
    return false
  }

  public func hash(into hasher: inout Hasher) {
    for i in 0..<16 {
      hasher.combine(byteAt(i))
    }
  }
}

extension Scru128Id: Codable {
  /// Encodes the object as a 25-digit canonical string representation.
  public func encode(to encoder: Encoder) throws {
    var container = encoder.singleValueContainer()
    try container.encode(description)
  }

  /// Decodes the object from a 25-digit canonical string representation or a 16-byte big-endian
  /// byte array.
  public init(from decoder: Decoder) throws {
    let container = try decoder.singleValueContainer()
    if let strValue = try? container.decode(String.self) {
      do {
        try self.init(parsing: strValue)
      } catch let err as ParseError {
        throw DecodingError.dataCorruptedError(in: container, debugDescription: err.description)
      }
    } else if let byteArray = try? container.decode([UInt8].self) {
      if byteArray.count == 16 {
        self.init(byteArray)
      } else if byteArray.allSatisfy({ $0 < 0x80 }) {
        let strValue = String(cString: byteArray + [0])
        do {
          try self.init(parsing: strValue)
        } catch is ParseError {
          throw DecodingError.dataCorruptedError(
            in: container, debugDescription: "could not parse byte array as Scru128Id")
        }
      } else {
        throw DecodingError.dataCorruptedError(
          in: container, debugDescription: "could not parse byte array as Scru128Id")
      }
    } else {
      throw DecodingError.dataCorruptedError(
        in: container, debugDescription: "expected string or byte array but found neither")
    }
  }
}

extension Scru128Id: Sendable {}
